mqtt_as/interfaces/__init__.py

```diff
@@ -0,0 +1,92 @@
+from uerrno import EINPROGRESS, ETIMEDOUT
+import usocket
+import uasyncio as asyncio
+
+
+async def _g():
+    pass
+
+
+_type_coro = type(_g())
+
+
+# If a callback is passed, run it and return.
+# If a coro is passed initiate it and return.
+# coros are passed by name i.e. not using function call syntax.
+def launch(func, tup_args):
+    res = func(*tup_args)
+    if isinstance(res, _type_coro):
+        res = asyncio.create_task(res)
+    return res
+
+
+class BaseInterface:
+    def __init__(self, socket=None):
+        # Legitimate errors while waiting on a socket. See uasyncio __init__.py open_connection().
+        self.BUSY_ERRORS = [EINPROGRESS, ETIMEDOUT]
+        self.socket = socket or usocket  # support for custom socket implementations
+        self._subs = []
+        self._state = None
+
+    async def connect(self):
+        """Serve connect request. Triggers callbacks if state changes"""
+        if await self._connect():
+            self._change_state(True)
+            return True
+        return False
```

Comment:
There should be a variable that prevents the method from being called when it's already running (e.g. ``self._is_connecting``).
Connecting can take up huge amounts of time and it doesn't make sense to have those running in parallel.
With the variable it's possible to just spawn connect tasks.

Reply:
I do agree actually. It is not a problem if only one project/library is controlling the interface but if multiple interfaces try it, then it becomes a problem. A 2nd interface calling connect should just wait for the result of the 1st call. 
However, it might be better to create a more advanced interface blueprint in order to keep this one as simple as possible and as close to the source as possible. The goal of the current files is to just separate mqtt_as from the hardware interface without adding features or complexity.
I do agree that more features and complexity is needed for a safe approach handling multiple libraries. So once the current changes/approach is approved, I should add such a blueprint (by making the wlan interfaces better).